lib/amazonka/src/Amazonka/Auth/SSO.hs

```diff
@@ -0,0 +1,126 @@
+import Amazonka.Data.Sensitive
+import Amazonka.Data.Time (Time (..))
+import Amazonka.Env (Env, Env' (..))
+import Amazonka.HTTP (retryRequest)
+import Amazonka.Lens ((^.))
+import Amazonka.Prelude
+import Amazonka.SSO.GetRoleCredentials as SSO
+import qualified Amazonka.SSO.Types as SSO (RoleCredentials (..))
+import Amazonka.Types
+import qualified Control.Exception as Exception
+import Control.Exception.Lens (handling_, _IOException)
+import Control.Monad.Trans.Resource (runResourceT)
+import Data.Aeson (FromJSON, decodeFileStrict)
+import qualified Data.Text as Text
+import qualified Data.Text.Encoding as Text
+import Data.Time.Clock.POSIX (posixSecondsToUTCTime)
+import qualified Network.HTTP.Client as Client
+
+data CachedAccessToken = CachedAccessToken
+  { startUrl :: Text,
+    region :: Region,
+    accessToken :: Sensitive Text,
+    expiresAt :: UTCTime
+  }
+  deriving stock (Show, Eq, Generic)
+  deriving anyclass (FromJSON)
+
+-- | Assume a role using an SSO Token
+--
+-- The user must have called @aws sso login@ and have appropriate @sso_@ values
+-- in their active profile, which are the arguments to this function. This
+-- function uses 'fetchAuthInBackground' to refresh the credentials as long as
+-- the token in the @sso/cache@ file is not expired. When it has, the user will
+-- need to @aws sso login@ again.
+--
+-- <https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-sso.html>
+fromSSO ::
+  (MonadIO m, Foldable withAuth) =>
+  FilePath ->
+  Region ->
+  Text ->
+  Text ->
+  Env' withAuth ->
+  m Env
+fromSSO cachedTokenFile ssoRegion accountId roleName env = do
+  auth <- liftIO $ fetchAuthInBackground getCredentials
+  pure $ env {envAuth = Identity auth}
+  where
+    getCredentials = do
+      CachedAccessToken {..} <- readCachedAccessToken cachedTokenFile
+
+      -- The Region you SSO through may differ from the Region you intend to
+      -- interact with after. The former is handled here, the latter is taken
+      -- care of later, in ConfigFile.
+      let ssoEnv = env {envRegion = ssoRegion}
+
+      resp <-
+        runResourceT $
+          sendUnsigned ssoEnv $
+            SSO.newGetRoleCredentials roleName accountId (fromSensitive accessToken)
+
+      let mCreds = do
+            rc <- resp ^. SSO.getRoleCredentialsResponse_roleCredentials
+            roleCredentialsToAuthEnv rc
+
+      case mCreds of
+        Nothing -> fail "sso:GetRoleWithCredentials returned no credentials."
+        Just c -> pure c
+
+sendUnsigned ::
+  ( MonadResource m,
+    AWSRequest a,
+    Foldable withAuth
+  ) =>
+  Env' withAuth ->
+  a ->
+  m (AWSResponse a)
+sendUnsigned env req = do
+  eResponse <- retryRequest env req
+  either (liftIO . Exception.throwIO) (pure . Client.responseBody) eResponse
+
+-- | Return the cached token file for a given @sso_start_url@
+--
+-- Matches
+-- <https://github.com/boto/botocore/blob/c02f3561f56085b8a3f98501d25b9857b916c10e/botocore/utils.py#L2596-L2597>
```

Comment:
```suggestion
-- [botocore](https://github.com/boto/botocore/blob/c02f3561f56085b8a3f98501d25b9857b916c10e/botocore/utils.py#L2596-L2597),
-- so that we find tokens produced by @aws sso login@.
```

Reply:
Whoa, when did Haddock start supporting markdown for links? I had to look it up to confirm but you're totally right.